File: leetcode/algorithms/matrix/flood-fill.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Flood Fill"
*/

function floodFill(image: number[][], sr: number, sc: number, newColor: number): number[][] {
    
    const currentColor: number = image[sr][sc];
    
    if(currentColor!==newColor){
        floodFillColor(image,sr,sc,newColor,currentColor);
    }
    
    return image;
    

};

function floodFillColor(image: number[][], r: number, c: number, newColor: number, currColor: number): void {
    if(r<0 || c<0 || r>=image.length || c>=image[0].length || image[r][c]!==currColor){
        return;
    }

    image[r][c]=newColor;

    if(r<image.length)floodFillColor(image,r+1,c,newColor,currColor);
    if(r>0)floodFillColor(image,r-1,c,newColor,currColor);
    if(c<image[0].length)floodFillColor(image,r,c+1,newColor,currColor);
    if(c>0)floodFillColor(image,r,c-1,newColor,currColor);

}

File: leetcode/algorithms/matrix/flood-fill.java
class Solution {
    private void floodFillImpl(int[][] image,int sr,int sc,int newColor,int oldColor){
        if((sr<0 || sr>image.length-1) || (sc<0 || sc>image[0].length-1) || image[sr][sc]!=oldColor){
            return;
        }
        
        image[sr][sc] = newColor;
        
        floodFillImpl(image,sr+1,sc,newColor,oldColor);
        floodFillImpl(image,sr,sc+1,newColor,oldColor);
        floodFillImpl(image,sr-1,sc,newColor,oldColor);
        floodFillImpl(image,sr,sc-1,newColor,oldColor);
        
    }
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if(newColor!=image[sr][sc]){
            floodFillImpl(image,sr,sc,newColor,image[sr][sc]);
        }
        return image;
    }
}